ethereum/spec/src/main/java/tech/pegasys/teku/spec/logic/common/helpers/BeaconStateAccessors.java

```diff
@@ -142,4 +153,39 @@ public Bytes32 getRandaoMix(BeaconState state, UInt64 epoch) {
     int index = epoch.mod(config.getEpochsPerHistoricalVector()).intValue();
     return state.getRandao_mixes().getElement(index);
   }
+
+  public int getBeaconProposerIndex(BeaconState state) {
+    return getBeaconProposerIndex(state, state.getSlot());
+  }
+
+  public int getBeaconProposerIndex(BeaconState state, UInt64 requestedSlot) {
+    validateStateCanCalculateProposerIndexAtSlot(state, requestedSlot);
+    return BeaconStateCache.getTransitionCaches(state)
+        .getBeaconProposerIndex()
+        .get(
+            requestedSlot,
+            slot -> {
+              UInt64 epoch = miscHelpers.computeEpochAtSlot(slot);
+              Bytes32 seed =
+                  Hash.sha2_256(
+                      Bytes.concatenate(
+                          getSeed(state, epoch, config.getDomainBeaconProposer()),
+                          uintToBytes(slot.longValue(), 8)));
+              List<Integer> indices = getActiveValidatorIndices(state, epoch);
+              return miscHelpers.computeProposerIndex(state, indices, seed);
+            });
+  }
+
+  private void validateStateCanCalculateProposerIndexAtSlot(
+      final BeaconState state, final UInt64 requestedSlot) {
+    UInt64 epoch = miscHelpers.computeEpochAtSlot(requestedSlot);
```

Comment:
nit:  May as well be final since the other variable here is final.
```suggestion
    final UInt64 epoch = miscHelpers.computeEpochAtSlot(requestedSlot);
```